File: src/main/java/com/example/domain/Task.java
package com.example.domain;

import lombok.AllArgsConstructor;
import lombok.Data;

import javax.persistence.*;
import java.util.HashMap;
import java.util.Map;

@Entity
@Table(name = "tasks")
@Data
@AllArgsConstructor
public class Task {
    public class IllegalStatusException extends RuntimeException {
        public IllegalStatusException() {
            super();
        }

        public IllegalStatusException(String message) {
            super(message);
        }
    }

    public Task() {
        this.status = Status.OPEN;
    }

    @Id
    @GeneratedValue
    private Integer id;

    private String title;

    private String description;

    @Enumerated(EnumType.STRING)
    private Status status;

    public void setStatusByString(String str) {
        Status status = Status.getStatus(str);
        if (status == null) {
            throw new IllegalStatusException();
        }

        this.status = status;
    }

    public enum Status {
        OPEN, CLOSE, CANCEL;

        private static final Map<String, Status> statusMap = new HashMap<>();

        static {
            for (Status status : Status.values()) {
                statusMap.put(status.name(), status);
            }
        }

        public static Status getStatus(String status) {
            return statusMap.get(status);
        }
    }
}

File: src/main/java/com/example/domain/Task.kt
package com.example.domain

import java.util.*
import javax.persistence.*

@Entity
@Table(name = "tasks")
data class Task(
        var title: String? = null,
        var description: String? = null,
        @Enumerated(EnumType.STRING) var status: Status? = Task.Status.OPEN
) {
    inner class IllegalStatusException : RuntimeException {
        constructor() : super() {
        }

        constructor(message: String) : super(message) {
        }
    }

    @Id
    @GeneratedValue
    var id: Int? = null

    constructor(id: Int, title: String, description: String, status: Status? = Task.Status.OPEN) : this(title, description, status) {
        this.id = id
    }

    enum class Status {
        OPEN, CLOSE, CANCEL;


        companion object {

            private val statusMap = HashMap<String, Status>()

            init {
                for (status in Status.values()) {
                    statusMap.put(status.name, status)
                }
            }

            fun getStatus(status: String?): Status? {
                if (status == null) {
                    return null
                }

                return statusMap[status]
            }
        }
    }
}
